client/allocrunner/taskrunner/vault_hook_test.go

```diff
@@ -3,9 +3,567 @@
+		name     string
+		setupReq func() (*interfaces.TaskPrestartRequest, error)
+	}{
+		{
+			name: "recover from secrets dir",
+			setupReq: func() (*interfaces.TaskPrestartRequest, error) {
+				// Write token to secrets dir.
+				secretsDirPath := t.TempDir()
+				err := os.WriteFile(filepath.Join(secretsDirPath, vaultTokenFile), []byte("much secret"), 0666)
+				if err != nil {
+					return nil, err
+				}
+
+				req := &interfaces.TaskPrestartRequest{
+					TaskEnv: taskenv.NewEmptyTaskEnv(),
+					TaskDir: &allocdir.TaskDir{
+						SecretsDir: secretsDirPath,
+						PrivateDir: t.TempDir(),
+					},
+				}
+				return req, nil
+			},
+		},
+		{
+			name: "recover from private dir",
+			setupReq: func() (*interfaces.TaskPrestartRequest, error) {
+				// Write token to private dir.
+				privateDirPath := t.TempDir()
+				err := os.WriteFile(filepath.Join(privateDirPath, vaultTokenFile), []byte("much secret"), 0666)
+				if err != nil {
+					return nil, err
+				}
+
+				req := &interfaces.TaskPrestartRequest{
+					TaskEnv: taskenv.NewEmptyTaskEnv(),
+					TaskDir: &allocdir.TaskDir{
+						SecretsDir: t.TempDir(),
+						PrivateDir: privateDirPath,
+					},
+				}
+				return req, nil
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			hook := setupTestHook(t, nil)
+
+			req, err := tc.setupReq()
+			must.NoError(t, err)
+			req.Task = hook.task
+
+			// Ensure Prestart() returns in a reasonable time.
+			ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+			t.Cleanup(cancel)
+
+			var resp interfaces.TaskPrestartResponse
+			err = hook.Prestart(ctx, req, &resp)
+			must.NoError(t, err)
+			must.NoError(t, ctx.Err())
+
+			// Verify token was recovered and not derived.
+			client := hook.client.(*vaultclient.MockVaultClient)
+			must.MapLen(t, 0, client.JWTTokens())
+			must.MapLen(t, 0, client.LegacyTokens())
+		})
+	}
+}
+
+func TestTaskRunner_VaultHook_deriveError(t *testing.T) {
+	ci.Parallel(t)
+
+	t.Run("unrecoverable error", func(t *testing.T) {
+		vaultClient, _ := vaultclient.NewMockVaultClient("")
+		mockVaultClient := vaultClient.(*vaultclient.MockVaultClient)
+
+		hook := setupTestHook(t, &vaultHookConfig{
+			clientFunc: func(string) (vaultclient.VaultClient, error) {
+				return mockVaultClient, nil
+			},
+		})
+		req := &interfaces.TaskPrestartRequest{
+			TaskEnv: taskenv.NewEmptyTaskEnv(),
+			TaskDir: &allocdir.TaskDir{
+				SecretsDir: t.TempDir(),
+				PrivateDir: t.TempDir(),
+			},
+			Task: hook.task,
+		}
+		var resp interfaces.TaskPrestartResponse
+
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		t.Cleanup(cancel)
+
+		// Set unrecoverable error.
+		mockVaultClient.SetDeriveTokenWithJWTFn(func(_ context.Context, _ vaultclient.JWTLoginRequest) (string, error) {
+			// Cancel the context to simulate the task being killed.
+			cancel()
+			return "", structs.NewRecoverableError(errors.New("unrecoverable test error"), false)
```

Comment:
Shouldn't this be an unrecoverable error?

Reply:
It's kind of a bad API 😅 

That magic boolean set to `false` is what makes this error unrecoverable. I think it would be better to have two functions `NewRecoverableError`and `NewUnrecoverableError` that abstract away the boolean.

Reply:
Gotcha. :+1: